pkg/migrate/cstor/pool.go

```diff
@@ -0,0 +1,431 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	replicaPatch = `{
+	"spec": {
+		"replicas": 0
+	}
+}`
+	cspNameLabel           = "cstorpool.openebs.io/name"
+	cspUIDLabel            = "cstorpool.openebs.io/uid"
+	cspHostnameAnnotation  = "cstorpool.openebs.io/hostname"
+	cspiNameLabel          = "cstorpoolinstance.openebs.io/name"
+	cspiUIDLabel           = "cstorpoolinstance.openebs.io/uid"
+	cspiHostnameAnnotation = "cstorpoolinstance.openebs.io/hostname"
+	spcFinalizer           = "storagepoolclaim.openebs.io/finalizer"
+	cspcFinalizer          = "cstorpoolcluster.openebs.io/finalizer"
+)
+
+// CSPCMigrator ...
+type CSPCMigrator struct {
+	// kubeclientset is a standard kubernetes clientset
+	KubeClientset kubernetes.Interface
+	// openebsclientset is a openebs custom resource package generated for custom API group.
+	OpenebsClientset openebsclientset.Interface
+	CSPCObj          *cstor.CStorPoolCluster
+	SPCObj           *apis.StoragePoolClaim
+	OpenebsNamespace string
+}
+
+// Migrate ...
+func (c *CSPCMigrator) Migrate(name, namespace string) error {
+	c.OpenebsNamespace = namespace
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return errors.Wrap(err, "error building kubeconfig")
+	}
+	c.KubeClientset, err = kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "error building kubernetes clientset")
+	}
+	c.OpenebsClientset, err = openebsclientset.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "error building openebs clientset")
+	}
+	err = c.migrate(name)
+	return err
+}
+
+// Pool migrates the pool from SPC schema to CSPC schema
+func (c *CSPCMigrator) migrate(spcName string) error {
+	var err error
+	var migrated bool
+	c.SPCObj, migrated, err = c.getSPCWithMigrationStatus(spcName)
+	if migrated {
+		klog.Infof("spc %s is already migrated to cspc", spcName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	err = c.validateSPC()
+	if err != nil {
+		return err
+	}
+	err = c.updateBDCLabels(spcName)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Creating equivalent cspc for spc %s", spcName)
+	c.CSPCObj, err = c.generateCSPC()
+	if err != nil {
+		return err
+	}
+	err = c.updateBDCOwnerRef()
+	if err != nil {
+		return err
+	}
+	// List all cspi created with reconcile off
+	cspiList, err := c.OpenebsClientset.CstorV1().
+		CStorPoolInstances(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + c.CSPCObj.Name,
+		})
+	if err != nil {
+		return err
+	}
+
+	// For each cspi perform the migration from csp that present was on
+	// node on which cspi came up.
+	for _, cspiItem := range cspiList.Items {
+		// Skip the migration if cspi is already in ONLINE state,
+		// which implies the migration is done and makes it idempotent
+		cspiItem := cspiItem // pin it
+		cspiObj := &cspiItem
+		if cspiObj.Status.Phase != "ONLINE" {
```

Comment:
This will cause the problem when the pod got restart after pool is successfully imported we will left with CVC having SPC information.